### PR TITLE
Modified mongodb-dev.rb to mongodb-dev that compiles on OSX 10.11

### PR DESCRIPTION
**Pull request by fjonath1**
I created custom mongodb-dev that disables -Werror by default, and backports some patches from newer version of mongodb.

### CODE REVIEW

File: mongodb-dev.rb [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 class MongodbDev < Formula
   homepage 'http://www.mongodb.org/'
-  url 'https://github.com/mongodb/mongo/archive/r2.5.4.tar.gz'
-  sha1 '0d8b8dad3b5909af2b9b7ba7ed22d5cb0a7cfd98'
+  url 'https://github.com/fjonath1/mongodb-ros-osx/archive/master.zip'
+  sha1 '8e045846213484ca5de64217daf401f0ffd72237'
```
Comment (wjwwood) [inline mongodb-dev.rb]: Please do patch files through Homebrew rather than pointing it to your fork.

File: mongodb-dev.rb [hunk 2]
```diff
@@ -12,11 +12,11 @@ class MongodbDev < Formula
   conflicts_with "mongodb", :because=>"installs the same binaries"
 
   def install
-    args =  ["--full", "--use-system-boost", "--prefix=#{prefix}"]
+    args =  ["--full", "--use-system-boost", "-j4", "--prefix=#{prefix}"]
```
Comment (wjwwood) [inline mongodb-dev.rb]: This is not necessary and is controller by Homebrew. https://github.com/Homebrew/homebrew/issues/3118

File: mongodb-dev.rb [hunk 3]
```diff
@@ -12,11 +12,11 @@ class MongodbDev < Formula
...
 
-    if ENV.compiler == :clang && MacOS.version >= :mavericks
+    if ENV.compiler == :clang
```
Comment (wjwwood) [inline mongodb-dev.rb]: Why did you remove the `>= :mavericks`?

File: mongodb-dev.rb [hunk 4]
```diff
@@ -12,11 +12,11 @@ class MongodbDev < Formula
...
         args << "--64"
-        args << "--libc++"
+        # args << "--libc++"
```
Comment (wjwwood) [inline mongodb-dev.rb]: Why was this removed?

**Comment by wjwwood** [comment]
Thanks for the pull request. I had a few comments which need to be resolved before I can merge, but I'd appreciate your time in iterating on them with me.

File: mongodb-dev.rb (hunk 1 above)
Comment (fjonath1) [inline mongodb-dev.rb]: Do you know instructions to do that?

File: mongodb-dev.rb (hunk 2 above)
Comment (fjonath1) [inline mongodb-dev.rb]: Scons doesn't do that automatically, Cmake does I believe. I verified it with the cpu usage.

File: mongodb-dev.rb (hunk 1 above)
Comment (wjwwood) [inline mongodb-dev.rb]: I'd recommend updating to the `mongodb` release you're basing off of, then for each of your custom patches you can use `git diff` (http://stackoverflow.com/questions/5159185/create-a-git-patch-from-the-changes-in-the-current-working-directory) to get your changes. If you've already made commits (for your additional changes off the `mongodb` release) then you can use `git format-patch` (http://stackoverflow.com/questions/6658313/generate-a-git-patch-for-a-specific-commit).

You can look at other Homebrew formula for examples of how to include patches in the formula. I recommend hosting the patch files on gist.github.com or put them in the data section at the end of the formula file.

Comment (wjwwood) [inline mongodb-dev.rb]: Also the Homebrew cookbook is a good place to look for patches: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Formula-Cookbook.md#patches

File: mongodb-dev.rb (hunk 2 above)
Comment (wjwwood) [inline mongodb-dev.rb]: I didn't notice it was scons and not cmake/make. That's fine then.

File: mongodb-dev.rb (hunk 3 above)
Comment (fjonath1) [inline mongodb-dev.rb]: Newer mongodb version has some patches to include el capitan.
If I just run it as it was, I ran into a problem due to brew/scons (not sure) think el capitan os version is lower than maverics. I don't really understand why, but this patch from mongodb26 seems fixed it.
```
  # Review this patch with each release.
  # This modifies the SConstruct file to include 10.10 and 10.11 as accepted build options.
  if MacOS.version >= :yosemite
    patch do
      url "https://gist.githubusercontent.com/asmarques/4e4e66121dbe46a08933/raw/d6753551ff3849bf0fb5ca87c0a96e1255f4254d/mongodb26-elcapitan.diff"
      sha256 "2f79588a8a15660908d8d071655541c9ab0ac425550b02dc454c861c75f86606"
    end
  end

```

File: mongodb-dev.rb (hunk 4 above)
Comment (fjonath1) [inline mongodb-dev.rb]: I cannot build it with that option, I recall there some error when building it with the libc++.
I'm trying again to recompile it with libc++ to see what the actual error is.

Comment (wjwwood) [inline mongodb-dev.rb]: Ok, I'm just worried about using this library compiled against `--stdlib=libstdc++` (I presume) with other libraries on OS X which are compiled against `--stdlib=libc++` by default.

**Comment by fjonath1** [comment]
@wjwwood I made a new version.

File: mongodb-dev.rb (hunk 4 above)
Comment (fjonath1) [inline mongodb-dev.rb]: Ok, I revert it with args << "--libc++" and it works.
I cannot remember why I run thru any problem before with that args.

**Comment by wjwwood** [comment]
Thanks, are you going to incorporate the patch from https://github.com/ros/homebrew-deps/pull/12#discussion_r53107460 or is that a patch to SCons itself?

A lot of these patches seem strange to me, almost as if they are reverts (a lot of removing checks that `this != nullptr` whereas I'd expect patches would add those. Maybe that's just a red herring, but could you give me a link to one of these changes upstream?

**Comment by fjonath1** [comment]
It seems to works fine now with the MacOS.version check. I don't know why I get problem before with the version check.

I think the scons patch will add MacOS 10.10 and 10.11 compatibility if we are running the build with args `--osx-version-min=10.10` or `--osx-version-min=10.11`. Since building the package with args `--osx-version-min=10.9` works just fine, it should not matter much.

The patches I had done for this != nullptr comes from several commits that I found in newer version (around 2.6 to 3.x). I can try to find it and get back to you.

**Comment by wjwwood** [comment]
Ok, I just want to make sure that you have the patches in the correct direction. They may be fine.

**Comment by fjonath1** [comment]
This is all I remembered.

dbclientcursor.h Fixes
https://jira.mongodb.org/browse/CXX-377
mathutil.h and mathutil.cc Fixes
https://jira.mongodb.org/browse/SERVER-15183
s2regioncoverer.cc (Forgot to add this fix, but it still builds fine because -Werror disabled)
https://jira.mongodb.org/browse/SERVER-15184
Flag --disable-warnings-as-errors error on new version (which basically disables -Werror on SConstruct)
https://opensource-jira.infn.it/jira/browse/CHSFRMWRK-24
scons patch things to add osx-min-version 10.10 and 10.11 (if we want to add that patch)
https://jira.mongodb.org/browse/CXX-648)

**Comment by wjwwood** [comment]
Cool, those look right. I'm just trying it out on my machine and if that goes well, I'll merge it. Thanks again!

**Comment by wjwwood** [comment]
On my machine I get a failure when it tries to find `boost`:

```
% brew install https://raw.githubusercontent.com/fjonath1/homebrew-deps/master/mongodb-dev.rb
######################################################################## 100.0%
==> Downloading https://github.com/mongodb/mongo/archive/r2.5.4.tar.gz
Already downloaded: /Library/Caches/Homebrew/mongodb-dev-2.5.4.tar.gz
==> Patching
patching file SConstruct
patching file src/mongo/client/dbclientcursor.cpp
patching file src/mongo/client/dbclientcursor.h
patching file src/mongo/db/geo/geoparser.cpp
patching file src/mongo/db/geo/hash.cpp
patching file src/mongo/db/geo/shapes.cpp
patching file src/mongo/db/pdfile.cpp
patching file src/mongo/shell/linenoise_utf8.h
patching file src/mongo/unittest/unittest.cpp
patching file src/mongo/util/concurrency/mutexdebugger.h
patching file src/mongo/util/md5main.cpp
patching file src/third_party/s2/util/math/mathutil.cc
patching file src/third_party/s2/util/math/mathutil.h
==> scons install --full --use-system-boost -j4 --prefix=/usr/local/Cellar/mongodb-dev/2.5.4 --64 --libc++ --osx-version-min=10.9
Last 15 lines from /Users/william/Library/Logs/Homebrew/mongodb-dev/01.scons:
Checking if C++ compiler "g++" is GCC... no
Checking if C++ compiler "g++" is clang... yes
Checking if C compiler "gcc" is clang... yes
Checking if C compiler supports -Wno-unused-local-typedefs... yes
Checking if C compiler supports -Wno-unused-function... yes
Checking if C compiler supports -Wno-unused-private-field... yes
Checking if C compiler supports -Wno-mismatched-tags... yes
Checking if C compiler supports -Wno-deprecated-declarations... yes
Checking if C compiler supports -Wno-tautological-constant-out-of-range-compare... yes
Checking if C compiler supports -Wno-unused-const-variable... yes
Checking if C compiler supports -mmacosx-version-min=10.9... yes
Checking if C++ compiler supports -stdlib=libc++... yes
Checking for __thread... yes
Checking for C++ header file boost/filesystem/operations.hpp... no
can't find boost headers

READ THIS: https://git.io/brew-troubleshooting
```

Here's my boost info:

```
% brew info boost
boost: stable 1.60.0 (bottled), HEAD
Collection of portable C++ source libraries
http://www.boost.org
/usr/local/Cellar/boost/1.57.0 (10,572 files, 414.7M)
  Poured from bottle
/usr/local/Cellar/boost/1.58.0 (10,718 files, 462.2M)
  Poured from bottle
/usr/local/Cellar/boost/1.59.0 (10,928 files, 429.2M)
  Poured from bottle
/usr/local/Cellar/boost/1.60.0_1 (11,139 files, 436.0M) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/boost.rb
==> Dependencies
Optional: icu4c ✔
==> Options
--c++11
	Build using C++11 mode
--universal
	Build a universal binary
--with-icu4c
	Build regexp engine with icu support
--with-mpi
	Build with MPI support
--without-single
	Disable building single-threading variant
--without-static
	Disable building static library variant
--HEAD
	Install HEAD version
```

I have a few versions installed, but the only one that is linked is `1.60.0_1`.

**Comment by wjwwood** [comment]
It looks like the header is in there:

```
% find /usr/local/Cellar/boost/1.60.0_1/include/boost | grep filesystem/operations
/usr/local/Cellar/boost/1.60.0_1/include/boost/filesystem/operations.hpp
```

**Comment by wjwwood** [comment]
Looking into it briefly it looks like scons is not getting the boost include directories added to the include path. I'm not sure what the "right" mechanism for this is, but it some how needs to get extended to include either `/usr/local/include` or `/usr/local/Cellar/boost/1.60.0_1/include`.

**Comment by ahundt** [comment]
Worked for me with 1.60.0_1 (I'm sitting right next to @fjonath1 )

**Comment by ahundt** [comment]
perhaps `brew unlink boost; brew link boost` will make it happier?

**Comment by wjwwood** [comment]
Unfortunately it does not. @ahundt how did you run the install (the exact line)?

**Comment by ahundt** [comment]
@wjwwood 
```
brew tap fjonath1/deps
brew install fjonath1/deps/mongodb-dev
```

**Comment by fjonath1** [comment]
I can run the installation with boost 1.60.0_1 too.
This is my boost info
```
Felixs-MacBook-Pro:~ FelixJonathan$ brew info boost
boost: stable 1.60.0 (bottled), HEAD [pinned at 1.59.0]
Collection of portable C++ source libraries
http://www.boost.org
/usr/local/Cellar/boost/1.59.0 (10,928 files, 429.2M)
  Poured from bottle
/usr/local/Cellar/boost/1.60.0_1 (11,139 files, 436.0M) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/boost.rb
==> Dependencies
Optional: icu4c ✔
==> Options
--c++11
	Build using C++11 mode
--universal
	Build a universal binary
--with-icu4c
	Build regexp engine with icu support
--with-mpi
	Build with MPI support
--without-single
	Disable building single-threading variant
--without-static
	Disable building static library variant
--HEAD
	Install HEAD version
```

```
% find /usr/local/Cellar/boost/1.60.0_1/include/boost | grep filesystem/operations
/usr/local/Cellar/boost/1.60.0_1/include/boost/filesystem/operations.hpp
```


The command I run and the result
```
Felixs-MacBook-Pro:~ FelixJonathan$ brew install fjonath1/deps/mongodb-dev
==> Installing mongodb-dev from fjonath1/deps
==> Downloading https://github.com/mongodb/mongo/archive/r2.5.4.tar.gz
Already downloaded: /Library/Caches/Homebrew/mongodb-dev-2.5.4.tar.gz
==> Patching
patching file SConstruct
patching file src/mongo/client/dbclientcursor.cpp
patching file src/mongo/client/dbclientcursor.h
patching file src/mongo/db/geo/geoparser.cpp
patching file src/mongo/db/geo/hash.cpp
patching file src/mongo/db/geo/shapes.cpp
patching file src/mongo/db/pdfile.cpp
patching file src/mongo/shell/linenoise_utf8.h
patching file src/mongo/unittest/unittest.cpp
patching file src/mongo/util/concurrency/mutexdebugger.h
patching file src/mongo/util/md5main.cpp
patching file src/third_party/s2/util/math/mathutil.cc
patching file src/third_party/s2/util/math/mathutil.h
==> scons install --full --use-system-boost -j4 --prefix=/usr/local/Cellar/mongo
Warning: This keg was marked linked already, continuing anyway
```

**Comment by ahundt** [comment]
my boost configuration:
```
boost: stable 1.60.0 (bottled), HEAD
Collection of portable C++ source libraries
http://www.boost.org
/usr/local/Cellar/boost/1.59.0 (10,851 files, 119.8M)
  Built from source with: --with-mpi --without-single --without-static --c++11
/usr/local/Cellar/boost/1.60.0_1 (11,059 files, 123.0M) *
  Built from source with: --with-mpi --without-single --without-static --c++11
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/boost.rb
==> Dependencies
Optional: icu4c ✔
==> Options
--c++11
	Build using C++11 mode
--universal
	Build a universal binary
--with-icu4c
	Build regexp engine with icu support
--with-mpi
	Build with MPI support
--without-single
	Disable building single-threading variant
--without-static
	Disable building static library variant
--HEAD
	Install HEAD version
```

**Comment by wjwwood** [comment]
Using it as a tap doesn't make a difference for me either... Let me investigate it some more and I'll let you know.

**Comment by wjwwood** [comment]
What versions of OS X are you guys running?

**Comment by fjonath1** [comment]
@ahundt and I are using OSX 10.11.3 .

**Comment by wjwwood** [comment]
I'm using 10.11.2, but on my build farm machine which is running 10.10.x it appears to be working.

**Comment by wjwwood** [comment]
Did you guys turn off SIP (https://en.wikipedia.org/wiki/System_Integrity_Protection)?

**Comment by fjonath1** [comment]
My SIP is disabled. I don't know about @ahundt .

**Comment by wjwwood** [comment]
That's likely why it works for you. I don't think it's an option to require everyone to disable SIP in order to use this. I was homing in on the root of the issue, but ran out of time to work on it. I'll try to take another look tomorrow.

**Comment by wjwwood** [comment]
I managed to get it working on my machine with this patch:

```diff
diff --git a/mongodb-dev.rb b/mongodb-dev.rb
index 6e4ed1c..3d950ff 100644
--- a/mongodb-dev.rb
+++ b/mongodb-dev.rb
@@ -15,11 +15,13 @@ class MongodbDev < Formula
 
   def install
     args =  ["--full", "--use-system-boost", "-j4", "--prefix=#{prefix}"]
+    boost = Formula.factory('boost').prefix
 
     if ENV.compiler == :clang && MacOS.version >= :mavericks
         args << "--64"
         args << "--libc++"
         args << "--osx-version-min=10.9"
+        args << "--extrapath=#{boost}"
     end
 
     system "scons", "install", *args

```

I'm not completely certain how SIP affects this (perhaps SIP is not the source), but it appears there is precedence for this kind of change: https://github.com/TylerBrock/homebrew/commit/ac46e3aaa3f6652b55255675e3334e66189e119c

Which I found through: https://github.com/Homebrew/homebrew/issues/24950

Can one of you try that patch out on top of this pr and let me know if it works for you too?

If so, I'll merge after you add that patch to this change set. Thanks for your patience.

**Comment by fjonath1** [comment]
The patch worked great.
Thank you for helping me merge this.

Note: Out of curiosity, I disabled SIP and run the original script, and it still compiles, so the boost failure might be related to something else which will be fixed by the boost patch.

**Comment by wjwwood** [comment]
Yeah, the cause may not be related to SIP.

**Comment by wjwwood** [comment]
Merged with my patch.

**Comment by ahundt** [comment]
SIP is enabled for me, so I'm fairly sure it was unrelated. Was probably something boost or include path related.
